File: offboard_ws/src/cpp_swarm/src/cpp_uav.cpp
#include <px4_msgs/msg/offboard_control_mode.hpp>
#include <px4_msgs/msg/trajectory_setpoint.hpp>
#include <px4_msgs/msg/vehicle_command.hpp>
#include <px4_msgs/msg/vehicle_control_mode.hpp>
#include <rclcpp/rclcpp.hpp>
#include <stdint.h>

#include <chrono>
#include <iostream>
#include <array>
#include <cmath>
#include <eigen3/Eigen/Eigen>

using namespace std::chrono;
using namespace std::chrono_literals;
using namespace px4_msgs::msg;
using namespace std;

class OffboardControl : public rclcpp::Node
{
public:
	OffboardControl() : Node("offboard_control")
	{

		offboard_control_mode_publisher_ = this->create_publisher<OffboardControlMode>("/fmu/in/offboard_control_mode", 10);
		trajectory_setpoint_publisher_ = this->create_publisher<TrajectorySetpoint>("/fmu/in/trajectory_setpoint", 10);
		vehicle_command_publisher_ = this->create_publisher<VehicleCommand>("/fmu/in/vehicle_command", 10);

		mission_state = 0;
        current_waypoint = {0.0,0.0,0.0};
		timer_mission = this->create_wall_timer(10s, std::bind(&OffboardControl::mission_cb, this));
        timer_offboard = this->create_wall_timer(100ms, std::bind(&OffboardControl::offboard_cb, this));
	}

	void arm();
	void disarm();

private:
	rclcpp::TimerBase::SharedPtr timer_mission;
    rclcpp::TimerBase::SharedPtr timer_offboard;

	rclcpp::Publisher<OffboardControlMode>::SharedPtr offboard_control_mode_publisher_;
	rclcpp::Publisher<TrajectorySetpoint>::SharedPtr trajectory_setpoint_publisher_;
	rclcpp::Publisher<VehicleCommand>::SharedPtr vehicle_command_publisher_;

	std::atomic<uint64_t> timestamp_;   //!< common synced timestamped

	uint64_t mission_state;   // counter to define mission as state machine
    Eigen::Matrix<double,3,1> current_waypoint;

	void publish_offboard_control_mode();
	void publish_trajectory_setpoint();
	void publish_vehicle_command(uint16_t command, float param1 = 0.0, float param2 = 0.0);
    void mission_cb();
    void offboard_cb();
	Eigen::Matrix<double,3,1> ENU2NED_vector_converter(Eigen::Matrix<double,3,1> vector);
};

/**
 * @brief Define a mission composed by a set of waypoints for the drone to reach
*/
void OffboardControl::mission_cb()
{
    if(mission_state == 0)
    {
        RCLCPP_INFO(this->get_logger(), "Mission started");
        /*  ENABLE OFFBOARD MODE

            Python implementation of Offboard Control
            param1 is set to 1 to enable the custom mode.
            param2 is set to 6 to indicate the offboard mode.
            Other options for param2 include:
            param2 = 1 (MANUAL)
            param2 = 2 (ALTCTL)
            param2 = 3 (POSCTL)
            param2 = 4 (AUTO)
            param2 = 5 (ACRO)
            param2 = 6 (OFFBOARD)
            param2 = 7 (STABILIZED)
            param2 = 8 (RATTITUDE)
        */
        this->publish_vehicle_command(VehicleCommand::VEHICLE_CMD_DO_SET_MODE, 1., 6.);
        mission_state = 1;
    }else if (mission_state == 1)
    {
        /*  ARMING THE VEHICLE AND REACHING THE FIRST WAYPOINT*/
        RCLCPP_INFO(this->get_logger(), "Arming");
		this->arm();
        RCLCPP_INFO(this->get_logger(), "Reaching First Waypoint");

        current_waypoint = {-2.5,-2.5,2.0};
		current_waypoint = ENU2NED_vector_converter(current_waypoint);
		mission_state = 2;
    }
}

/**
 * @brief Define a mission composed by a set of waypoints for the drone to reach
*/
void OffboardControl::offboard_cb()
{
	publish_offboard_control_mode();
	publish_trajectory_setpoint();
} 

Eigen::Matrix<double,3,1> OffboardControl::ENU2NED_vector_converter(Eigen::Matrix<double,3,1> vector)
{
	// Rotation around Z-axis of 90° Degrees
	Eigen::Matrix<double,3,3> rotZ { {cos(M_PI/2), -sin(M_PI/2), 0},
									{sin(M_PI/2), cos(M_PI/2), 0}, 
									{0,0,1}};

	vector = rotZ*vector;

	// Rotation around X-axis of 180° Degrees
	Eigen::Matrix<double,3,3> rotX { {1,0,0}, 
									{0, cos(M_PI), -sin(M_PI)},
									{0, sin(M_PI), cos(M_PI)}};

	vector = rotX*vector;

	return vector;
}

/**
 * @brief Send a command to Arm the vehicle
 */
void OffboardControl::arm()
{
	publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0);

	RCLCPP_INFO(this->get_logger(), "Arm command send");
}

/**
 * @brief Send a command to Disarm the vehicle
 */
void OffboardControl::disarm()
{
	publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0);

	RCLCPP_INFO(this->get_logger(), "Disarm command send");
}

/**
 * @brief Publish the offboard control mode.
 *        For this example, only position and altitude controls are active.
 */
void OffboardControl::publish_offboard_control_mode()
{
	OffboardControlMode msg{};
	msg.position = true;
	msg.velocity = false;
	msg.acceleration = false;
	msg.attitude = false;
	msg.body_rate = false;
	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
	offboard_control_mode_publisher_->publish(msg);
}

/**
 * @brief Publish a trajectory setpoint
 *        For this example, it sends a trajectory setpoint to make the
 *        vehicle hover at 5 meters with a yaw angle of 180 degrees.
 */
void OffboardControl::publish_trajectory_setpoint()
{
	TrajectorySetpoint msg{};
	std::array<float, 3UL> floatWaypoint;
	for (int i = 0; i < 3; ++i) {
        floatWaypoint[i] = static_cast<float>(current_waypoint(i));
    }

	msg.position = floatWaypoint;
	// The drone starts with a yaw of pi/2
	msg.yaw = M_PI/2; // [-PI:PI]
	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
	trajectory_setpoint_publisher_->publish(msg);
}

/**
 * @brief Publish vehicle commands
 * @param command   Command code (matches VehicleCommand and MAVLink MAV_CMD codes)
 * @param param1    Command parameter 1
 * @param param2    Command parameter 2
 */
void OffboardControl::publish_vehicle_command(uint16_t command, float param1, float param2)
{
	VehicleCommand msg{};
	msg.param1 = param1;
	msg.param2 = param2;
	msg.command = command;
	msg.target_system = 1;
	msg.target_component = 1;
	msg.source_system = 1;
	msg.source_component = 1;
	msg.from_external = true;
	msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
	vehicle_command_publisher_->publish(msg);
}

int main(int argc, char *argv[])
{
	std::cout << "Starting offboard control node..." << std::endl;
	setvbuf(stdout, NULL, _IONBF, BUFSIZ);
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<OffboardControl>());

	rclcpp::shutdown();
	return 0;
}

File: offboard_ws/src/cpp_swarm/src/cpp_uav.py
#!/bin/python3

"""
SCRIPT THAT PERFORMS COVERAGE PATH PLANNING OF 1 DRONE
"""

import rclpy
from rclpy.node import Node
from rclpy.clock import Clock
import numpy as np
from enum import Enum

from px4_msgs.msg import OffboardControlMode
from px4_msgs.msg import TrajectorySetpoint
from px4_msgs.msg import VehicleCommand
from px4_msgs.msg import VehicleStatus

from rclpy.qos import (
    QoSProfile, 
    QoSReliabilityPolicy, 
    QoSHistoryPolicy)

MAP_WIDTH = 3
MAP_HEIGHT = 3
CELL_UNIT = 1
SUBUNIT_CELL = 0.5

class Dir_neighbours(Enum):
    POS_X = 1, 
    NEG_Y = 2, 
    NEG_X = 3, 
    POS_Y = 4

class Dir_subcells(Enum):
    TOP_RIGHT = 1, 
    TOP_LEFT = 2, 
    BOTTOM_LEFT = 3, 
    BOTTOM_RIGHT = 4

class SubCell:
    def __init__(self, x, y):
        self.x = x
        self.y = y
    
    def display(self):
         print("X= "+ str(self.x) + " Y = " + str(self.y))

    def compareSubCells(self, sub1, sub2):
        if (sub1.x == sub2.x and sub1.y == sub1.y):
            return True
        else:
            return False

class SubCellGroup(SubCell):
    
    def __init__(self, parent_x=0, parent_y=0):
        
        self.parent_x = parent_x
        self.parent_y = parent_y

        self.subcells[Dir_subcells.TOP_RIGHT].x=parent_x+SUBUNIT_CELL/2
        self.subcells[Dir_subcells.TOP_RIGHT].y=parent_y-SUBUNIT_CELL/2

        self.subcells[Dir_subcells.TOP_LEFT].x=parent_x-SUBUNIT_CELL/2
        self.subcells[Dir_subcells.TOP_LEFT].y=parent_y-SUBUNIT_CELL/2

        self.subcells[Dir_subcells.BOTTOM_LEFT].x=parent_x-SUBUNIT_CELL/2
        self.subcells[Dir_subcells.BOTTOM_LEFT].y=parent_y+SUBUNIT_CELL/2

        self.subcells[Dir_subcells.BOTTOM_RIGHT].x=parent_x+SUBUNIT_CELL/2
        self.subcells[Dir_subcells.BOTTOM_RIGHT].y=parent_y+SUBUNIT_CELL/2
        

class OffboardControl(Node):

    def __init__(self):
        super().__init__('OffboardControl')
        self.offboard_control_mode_publisher_ = self.create_publisher(OffboardControlMode,"/fmu/in/offboard_control_mode", 10)
        self.trajectory_setpoint_publisher_ = self.create_publisher(TrajectorySetpoint,"/fmu/in/trajectory_setpoint", 10)
        self.vehicle_command_publisher_ = self.create_publisher(VehicleCommand,"/fmu/in/vehicle_command", 10)
        self.vehicle_status_publisher_ = self.create_publisher(VehicleStatus,"/fmu/in/vehicle_status", 10)

        # Mission organized as a state machine
        self.mission_state = 0
        # Starting value of the waypoint updated within each iteration 
        # of the state machine inside timer_offbaord callback
        self.current_waypoint = [0.0, 0.0, 0.0]

        # Timers
        self.timer_offboard = self.create_timer(0.1, self.timer_offboard_cb)
        self.timer_mission = self.create_timer(10, self.mission)

        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1
        )

    def mission(self):

        if self.mission_state == 0:

            self.get_logger().info("Mission started")
            
            """ Arming and takeoff to 1st waypoint """
            """
            Python implementation of Offboard Control
            param1 is set to 1 to enable the custom mode.
            param2 is set to 6 to indicate the offboard mode.
            Other options for param2 include:
            param2 = 1 (MANUAL)
            param2 = 2 (ALTCTL)
            param2 = 3 (POSCTL)
            param2 = 4 (AUTO)
            param2 = 5 (ACRO)
            param2 = 6 (OFFBOARD)
            param2 = 7 (STABILIZED)
            param2 = 8 (RATTITUDE)
            """
            self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_DO_SET_MODE, 1., 6.)
            self.mission_state = 1

        elif self.mission_state == 1:    
            self.arm()
            self.get_logger().info("Vehicle armed")
          
            self.get_logger().info("First waypoint")
            self.current_waypoint = np.asfarray([-2.5, -2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 2

        elif self.mission_state == 2:
            """Waypoint 2"""
            self.get_logger().info("Second waypoint")
            # Imposta secondo waypoint, 
            self.current_waypoint = np.asfarray([-2.5, 2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            
            self.mission_state = 3
        
        elif self.mission_state == 3:
            """Waypoint 2"""
            self.get_logger().info("Third waypoint")
            # Imposta secondo waypoint
            self.current_waypoint = np.asfarray([2.5, -2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 4
        
        elif self.mission_state == 4:
            """Waypoint 2"""
            self.get_logger().info("Forth waypoint")
            # Imposta secondo waypoint
            self.current_waypoint = np.asfarray([2.5, 2.5, 2], dtype = np.float32)
            self.ENU2NED_vector_converter()
            print(self.current_waypoint)
            self.mission_state = 5

        elif self.mission_state == 5:
            """Landing"""
            self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_NAV_LAND)
            self.mission_state = 6
        
        elif self.mission_state == 6:
            self.timer_offboard.cancel()
            self.get_logger().info("Mission finished")
            self.timer_mission.cancel()
            exit()        
    
    def timer_offboard_cb(self):
        self.publish_offboard_control_mode()
        self.publish_trajectory_setpoint()

    # Arm the vehicle
    def arm(self):
        self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0)
        self.get_logger().info("Arm command send")

    # Disarm the vehicle
    # def disarm(self):
    #   self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0)
    #   self.get_logger().info("Disarm command send")


    ''' Publish the offboard control mode.
    For this example, only position and altitude controls are active.
    '''

    def publish_offboard_control_mode(self):
        msg = OffboardControlMode()
        msg.position = True
        msg.velocity = False
        msg.acceleration = False
        msg.attitude = False
        msg.body_rate = False
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
        self.offboard_control_mode_publisher_.publish(msg)

    '''
    Publish a trajectory setpoint
    For this example, it sends a trajectory setpoint to make the
    vehicle hover at 5 meters with a yaw angle of 180 degrees.
    '''

    def publish_trajectory_setpoint(self):
        msg = TrajectorySetpoint()
        msg.position = self.current_waypoint

        # The drone starts with a yaw of pi/2
        msg.yaw = np.pi/2  # [-PI:PI]
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
            
        self.trajectory_setpoint_publisher_.publish(msg)

    '''
    Publish vehicle commands
        command   Command code (matches VehicleCommand and MAVLink MAV_CMD codes)
        param1    Command parameter 1 as defined by MAVLink uint16 VEHICLE_CMD enum
        param2    Command parameter 2 as defined by MAVLink uint16 VEHICLE_CMD enum
        '''
    def publish_vehicle_command(self, command, param1=0.0, param2=0.0):
        msg = VehicleCommand()
        msg.param1 = param1
        msg.param2 = param2
        msg.command = command  # command ID
        msg.target_system = 1  # system which should execute the command
        msg.target_component = 1  # component which should execute the command, 0 for all components
        msg.source_system = 1  # system sending the command
        msg.source_component = 1  # component sending the command
        msg.from_external = True
        msg.timestamp = int(Clock().now().nanoseconds / 1000) # time in microseconds
        self.vehicle_command_publisher_.publish(msg)

    def ENU2NED_vector_converter(self):

        # Rotation around Z-axis of 90° Degrees
        rotZ = np.asfarray(np.array([[np.cos(np.pi/2), -np.sin(np.pi/2), 0],
                                    [np.sin(np.pi/2), np.cos(np.pi/2), 0], 
                                    [0, 0, 1]]), dtype = np.float32)
        self.current_waypoint = np.dot(rotZ,self.current_waypoint)

        # Rotation around X-axis of 180° Degrees
        rotX = np.asfarray(np.array([[1, 0, 0],
                                    [0, np.cos(np.pi), -np.sin(np.pi)], 
                                    [0, np.sin(np.pi), np.cos(np.pi)]]), dtype = np.float32)
        self.current_waypoint = np.dot(rotX,self.current_waypoint)
    
class Graph:
    def __init__(self, vertex=0):
        self.V = vertex
        self.graph = []
 
    def add_edge(self, u, v, w):
        self.graph.append([u, v, w])
 
 
    def search(self, parent, i):
        if parent[i] == i:
            return i
        return self.search(parent, parent[i])
 
    def apply_union(self, parent, rank, x, y):
        xroot = self.search(parent, x)
        yroot = self.search(parent, y)
        if rank[xroot] < rank[yroot]:
            parent[xroot] = yroot
        elif rank[xroot] > rank[yroot]:
            parent[yroot] = xroot
        else:
            parent[yroot] = xroot
            rank[xroot] += 1
 
  
    def kruskal(self):
        result = []
        i, e = 0, 0
 
        # Sorts the graph based on the weight key (item = [u,v,w] so item[2] is the weight)
        # The sorting is in ascending order so the first edge is of minimum weight
        self.graph = sorted(self.graph, key=lambda item: item[2])
        parent = []
        rank = []

        # Every vertice is initialized as parent of itself and with rank 0
        for node in range(self.V):
            parent.append(node)
            rank.append(0)

        while e < self.V - 1:
            u, v, w = self.graph[i]
            i = i + 1
            # Find the parent of u
            x = self.search(parent, u)

            # Find the parent of v
            y = self.search(parent, v)

            # If they have same parent means that the edge forms a cycle
            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.apply_union(parent, rank, x, y)

        for u, v, weight in result:
            print("Edge:",u, v,end =" ")
            print("-",weight)

        return result
   
class GridMap(Graph):

    def __init__(self, data=[], width=0, height=0, resolution=0,
                 center_x=0, center_y=0, origin=[]):
        
        """ Parameters that define the class:
        :data: vector defining if each cell is either 0 (free) or 100 (occupied)
        :param width: number of grid for width
        :param height: number of grid for height
        :param resolution: grid resolution [m]
        :param center_x: center of the grid in drone coords x position [m]
        :param center_y: center of the grid in drone coords y position [m]
        :origin: position of the first cell in drone coords
        """

        self.data = data
        self.width = width
        self.height = height
        self.resolution = resolution
        self.center_x = center_x
        self.center_y = center_y
        self.origin = origin

        self.n_data = self.width * self.height

    """ Defines the grid map area to cover"""
    def area_definition(self, x, y):
        resolution = 2.0 # meters

        # calculate dimensions of the squared surface
        # x_max = int((max(x) - min(x))/resolution)
        # y_max = int((max(y) - min(y))/resolution)

        x_max = 3
        y_max = 3

        data = []
        cell_free = 0
        for i in range(y_max):
            for j in range(x_max):
                # for now all cells are free to cover
                data.append(cell_free)
        
        width = x_max
        height = y_max
        center_x = 0
        center_y = 0
        origin = [min(x),min(y)]
        # TO DO: valutare se serve di memorizzare anche per ogni cella che posizine su i,j è
        # insieme al numero di cella
        gridmap = GridMap(data, width, height, resolution, center_x, center_y, origin)

        return gridmap

    """ Divide the grid map area to cover with multiple CPS"""
    def area_division(self, grid_map, uav_pose):
    
        # convert swarm pose to grid position (i,j)
        # (i,j) are the coordinates of the position of the drone inside the grid
        print(grid_map.width, grid_map.height, grid_map.origin[:])
        i = (grid_map.width-1) - int((uav_pose[0] - grid_map.origin[0])/grid_map.resolution)
        j = (grid_map.height-1) - int((uav_pose[1] - grid_map.origin[1])/grid_map.resolution)

        return i, j
    
    # TO DO: retrieve the path to the drone as a sequence of waypoints using a service
    # Kruskal algorithm taken from https://www.pythonpool.com/kruskals-algorithm-python/
    def mst_generation(self):

        rows = self.width
        cols = self.height
        
        num_nodes = 9
        edges = Graph(num_nodes)

        # Define the nodes based on the middle point of each gridmap
        for i in range(rows) :
            for j in range(cols) :

                # Scanning the grid map as a matrix but stored in a vector
                # create all the edges that connect adjancent vertices
                if (self.data[i*cols+j] == 0):
                
                    if (i>0 and self.data[(i-1)*cols+j] == 0):
                        edges.add_edge(i*cols+j, (i-1)*cols+j, 1)
                
                    if (i<rows-1 and self.data[(i+1)*cols+j] == 0):
                        edges.add_edge(i*cols+j, (i+1)*cols+j, 1)
                
                    if (j>0 and self.data[i*cols+j-1] == 0):
                        edges.add_edge(i*cols+j, i*cols+j-1, 1)
                
                    if (j<cols-1 and self.data[i*cols+j+1] == 0):
                        edges.add_edge(i*cols+j, i*cols+j+1, 1)
    
        mst = edges.kruskal() 
        print(edges.graph[:])
        return mst         


def main(args=None):
    # rclpy.init(args=args)
    # print("Starting Offboard Mission...\n")
    # offboard_control = OffboardControl()
    # rclpy.spin(offboard_control)

    # Edges of the squared surface that needs to be covered
    # Defined in coordinates w.r.t gazebo center
    x = [1, 1, -1, -1]
    y = [1, -1, -1, 1]

    G = GridMap()
    G = G.area_definition(x,y)
    mst = G.mst_generation()

    # offboard_control.destroy_node()
    # rclpy.shutdown()
    # print("Clean exit")
    # exit(0)

if __name__ == '__main__':
    main()
    
    
